### PR TITLE
Add to the list of edge cases for not zapping result nodes.

### PR DESCRIPTION
**Pull request by berlin-ab**
No longer zap a result node if the plan's target list and the
lefttree's target list do not match. We need the result node so it can
feed the correct target list to Append.

Thanks for the suggested solution by: 
Co-authored-by: Richard Guo <riguo@pivotal.io>

### CODE REVIEW

File: src/backend/cdb/cdbmutate.c [hunk 1]
```diff
@@ -2697,6 +2697,12 @@ void assign_plannode_id(PlannedStmt *stmt)
 	}
 }
 
+static bool
+targetlist_matches_lefttree_targetlist(Plan plan)
```
Comment (guofengrichard) [inline src/backend/cdb/cdbmutate.c]: PostgreSQL uses function `tlist_same_exprs()` to check whether two target lists contain the same expressions. It is basically does the same thing with your function. How about we leverage on PostgreSQL's function?

Comment (berlin-ab) [reply src/backend/cdb/cdbmutate.c]: Ah, thanks. I looked around for a function that did this while thinking of a place to store this function. I didn't find anything, but it sounds like this is it.

**Comment by berlin-ab** [comment]
> With this PR, I suspect there are other test cases that originally have zapped the Result nodes, but now retains them. We need to fix these test cases.

Hopefully there are no test cases to fix and the tests are testing the externally visible behavior, but yes, I'll fix the plan output tests if any fail.

File: src/backend/cdb/cdbmutate.c (hunk 1 above)
Comment (berlin-ab) [reply src/backend/cdb/cdbmutate.c]: Doing a quick search for `tlist_same_exprs` turned up no results. Looks like we don't yet have that function.

Comment (guofengrichard) [reply src/backend/cdb/cdbmutate.c]: Oh, we don't have this function in 5X. It was introduced in PostgreSQL 9.3.

**Comment by guofengrichard** [comment]
Function `tlist_same_exprs ()` only compares `TargetEntry->expr` when checking two target lists. The reason is as explained in the comment:
```
 * However, we can ignore the TargetEntry attributes resname, ressortgroupref,
 * resorigtbl, resorigcol, and resjunk, because those are only labelings that
 * don't affect the row values computed by the node.  (Moreover, if we didn't
 * ignore them, we'd frequently fail to make the desired optimization, since
 * the planner tends to not bother to make resname etc. valid in intermediate
 * plan nodes.)
```
So, it is not the best approach to use `equal()` directly here to compare two target lists. We'd better follow on what `tlist_same_exprs ()` does.

**Comment by berlin-ab** [comment]
> Function `tlist_same_exprs ()` only compares `TargetEntry->expr` when checking two target lists. The reason is as explained in the comment:
> 
> ```
>  * However, we can ignore the TargetEntry attributes resname, ressortgroupref,
>  * resorigtbl, resorigcol, and resjunk, because those are only labelings that
>  * don't affect the row values computed by the node.  (Moreover, if we didn't
>  * ignore them, we'd frequently fail to make the desired optimization, since
>  * the planner tends to not bother to make resname etc. valid in intermediate
>  * plan nodes.)
> ```
> So, it is not the best approach to use `equal()` directly here to compare two target lists. We'd better follow on what `tlist_same_exprs ()` does.

Thanks for the deep research. I've backported the function and put it to use.

File: src/test/regress/sql/limited_select_list_with_where_clause.sql [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+		 end('2012-03-01')
+		 every(interval '1 month')
+       );
+
+insert into foo values(2, '2012-01-01', '2012-01-01');
+
+-- when selecting a small select list
+-- and using more columns in the where clause
+-- then the executed plan should succeed.
+select a from foo where b>='2012-01-01' and c=b;
```
Comment (paul-guo-) [inline src/test/regress/sql/limited_select_list_with_where_clause.sql]: `select b from foo where b>='2012-01-01' and c=b;` does not panic without the patch. And the targetlist of append and result for the query are different. Chatted with Richard, that means there might be further improvement room for this patch, i.e. no need to eliminate for all the unequal targetlist case.

File: src/test/regress/greenplum_schedule [hunk 3]
```diff
@@ -122,7 +122,7 @@ test: gporca_faults
  
 test: aggregate_with_groupingsets 
 
-test: nested_case_null sort bb_mpph createdb_large_table
+test: nested_case_null sort bb_mpph createdb_large_table limited_select_list_with_where_clause
```
Comment (paul-guo-) [inline src/test/regress/greenplum_schedule]: I strongly disagree we create a new test file for such a corner case. This might slightly slow down testing in theory.

File: src/include/optimizer/tlist.h [hunk 4]
```diff
@@ -31,6 +31,9 @@ extern TargetEntry *tlist_member_ignoring_RelabelType(Expr *expr, List *targetli
 extern List *flatten_tlist(List *tlist);
 extern List *add_to_flat_tlist(List *tlist, List *vars, bool resjunk);
 
+// Added to Greenplum while tracking Postgres REL9_4_STABLE
```
Comment (paul-guo-) [inline src/include/optimizer/tlist.h]: Please avoid '//'. This was from c++ although works with many popular compilers (I guess) but it is not good.

File: src/test/regress/sql/limited_select_list_with_where_clause.sql [hunk 5]
```diff
@@ -0,0 +1,15 @@
+-- Given a partitioned table with data
+create table foo(a numeric, b date, c date)
```
Comment (paul-guo-) [inline src/test/regress/sql/limited_select_list_with_where_clause.sql]: Please avoid using very common table names like foo, bar, etc. This is easy to lead to conflict. Also please drop them finally to benefit pg_upgrade testing.

File: src/test/regress/greenplum_schedule (hunk 3 above)
Comment (berlin-ab) [reply src/test/regress/greenplum_schedule]: What's your rationale for it being slower? 

I created a new test file because it significantly speeds up iteration time while working on a test file. Sorting through existing tests while trying to add a new test to an existing `sql/` and `expected/` file slows down development.

File: src/include/optimizer/tlist.h (hunk 4 above)
Comment (berlin-ab) [reply src/include/optimizer/tlist.h]: The other comments in this file are using `//`, so I kept consistency.

File: src/test/regress/sql/limited_select_list_with_where_clause.sql (hunk 5 above)
Comment (berlin-ab) [reply src/test/regress/sql/limited_select_list_with_where_clause.sql]: I thought I had changed that. Will update.

File: src/test/regress/sql/limited_select_list_with_where_clause.sql (hunk 2 above)
Comment (berlin-ab) [reply src/test/regress/sql/limited_select_list_with_where_clause.sql]: I propose we get this patch in that solves the crash, and put a feature request in to further optimize.

File: src/test/regress/greenplum_schedule (hunk 3 above)
Comment (paul-guo-) [reply src/test/regress/greenplum_schedule]: A new small file introduces overhead due to various things (new file handling, random io), but admittedly yeah parallel & diff might affect also, so up to you.
